src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2525,39 +2525,97 @@ end
 
 function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.invpermute!!)},
                               df::AbstractDataFrame, p::AbstractVector{<:Integer})
-    toskip = Set{Int}()
+    nrow(df) != length(p) &&
+        throw(DimensionMismatch("Permutation does not have a correct length " *
+                                "(expected $(nrow(df)) but got $(length(p)))"))
+    
+    cp = _compile_permutation!(Base.copymutable(p))
+
+    isempty(cp) && return df
+
+    if fun === Base.invpermute!! 
+        reverse!(@view cp[1:end-1])
+    end
+
     seen_cols = IdDict{Any, Nothing}()
     for (i, col) in enumerate(eachcol(df))
-        if haskey(seen_cols, col)
-            push!(toskip, i)
-        else
+        if !haskey(seen_cols, col)
             seen_cols[col] = nothing
-        end
-        # p might be a column of df so we make sure we unalias
-        if col === p
-            p = copy(p)
+            _cycle_permute!(col, cp)
         end
     end
+    
+    return df
+end
 
-    pp = similar(p)
+# convert a classical permutation to zero terminated cycle 
+# notation, zeroing the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    firstindex(p) == 1 || 
+        throw(ArgumentError("Permutation vectors must have 1-based indexing"))
+    # this length is sufficient because we do not record 1-cycles,
+    # so the worst case is all 2-cycles. One extra element gives the
+    # algorithm leeway to defer error detection without unsafe reads.
+    # trace _compile_permutation!([3,3,1]) for example.
+    out = similar(p, 3 * length(p) ÷ 2 + 1)
+    out_len = 0
+    start = 0
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start + 1)
+        start isa Int || throw(ArgumentError("Passed vector p is not a valid permutation"))
+        last_k = p[start]
+        count -= 1
+        last_k == start && continue
+        out_len += 1
+        out[out_len] = last_k
+        p[start] = 0
+        start <= last_k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

Comment:
```suggestion
        start < last_k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

Reply:
@LilithHafner - I have added a suggestion that my code analysis shows that are valid. Can you please confirm before I merge the PR that:
1. we do not need to set `p[start]` to `0` as we never go back to check the value of `p[start]` in the algorithm (actually you take advantage of this fact already in line `last_k == start && continue` if I understand your code correctly).
2. `last_k` cannot be equal to `start` here.

This change does not change the result of the code, but I just wanted to ensure that I understand your code correctly. The only benefit is that the condition would be a bit tighter + we would have some small performance gain (most visible if permutation consists only of 2-cycles).

Reply:
@LilithHafner so the `p[start] = 0` part is needed here because of `k = out[out_len] = p[k]`. Right? This makes sense.

Reply:
1. No. Without any other changes, removing `p[start] = 0` will cause problems because cycle termination detection is done by checking that `k == 0`, so if the starting element is not zeroed out before, it will be zeroed out as the last element of the cycle, thus double counting it.

2. Yes. If `last_k == start`, we've got a problem slightly more serious than the passed vector is not a permutation, but a stricter condition is nice.

Reply:
> will cause problems because cycle termination detection is done by checking that `k == 0`

This is what I meant. Thank you for confirming.

Reply:
Oops! I was out of sync and posted before I saw your post :) I'm glad we both drew the same conclusion.